missed updates from pr 710

mystiberry-okta opened the pull request:
## Description:
- Describe your changes. This will most likely be a copy-paste of your commit
  message(s), which should be descriptive and informative.

## Type of Pull Request:
<!--- What types of PR is this? Put an `x` in all the boxes that apply: -->
- [x] Content (Documentation updates or typo-fixes)
- [ ] Blog Post
- [ ] Functional (CSS changes, JS updates, or layout modifications)

### Resolves:
* [OKTA-121638](https://oktainc.atlassian.net/browse/OKTA-121638)

## How Has This Been Tested?
- [x] I have built this locally and verified that it does not break existing functionality.
- [ ] For functional changes, I have tested against Chrome, Firefox, and Safari, and mobile.
- [ ] For functional changes, I have added tests.

## Primary Reviewers:
@yuliu-okta optionally @cwu-okta and anyone in Fed team!

mystiberry-okta commented:
@yuliu-okta You'll notice the two tables--I created them to consolidate all the scattered info about grant_type+scope. Did I get everything? Thanks!

yuliu-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -188,46 +187,46 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+| scope         | Optional.          [ Different scopes and tokens](#response-parameters) are returned depending on the values of `scope` and `grant_type`.                                                                                                                                                                                         | String |
+| redirect_uri  | Required if *grant_type* is `authorization_code`. Specifies the callback location where the authorization was sent; must match what is preregistered in Okta for this client.                                                                                                                                            | String |
+| code_verifier | Required if *grant_type* is `authorization_code`  and `code_challenge` was specified in the original `/authorize` request. The code verifier of               [PKCE](#request-parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request. | String |
+| client_id     | Required if client credentials are not provided in the Authorization header. This is used in conjunction with the *client_secret* parameter to authenticate the client application.                                                                                                                                      | String |
```
Required if the client has a secret and client credentials are not provided in the Authorization header.

mystiberry-okta replied:
done.

yuliu-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -188,46 +187,46 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+
+| Requested grant type | Requested scope                                                            | Tokens in the response                                                       |
+|:---------------------|:---------------------------------------------------------------------------|:----------------------------------------------------------------------------------|
+| authorization_code   | None                                                                       | Access Token, Refresh Token, ID Token                                             |
```
If the grant type is authorization_code, there is no scope requested in the token endpoint, it will re-use the scopes requested in the authorize endpoint.
And for all grant type, it will generate the access token by default. For authorization_code, refresh_token and password grant type, if the client also requests offline_access scope, then it will also generate refresh token. For authorization_code and password, if the client also requests openid scope, then it will also generate id token.

mystiberry-okta replied:
Great! These details weren't clear before, that's why I put them all in a table. THANK YOU!

yuliu-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -254,7 +253,7 @@ For web and native application types, an additional process is required:
 | invalid_client          | The specified client ID wasn't found or authentication failed.                                                                                                                                                                 |
 | invalid_request         | The request structure was invalid. E.g. the basic authentication header was malformed, or both header and form parameters were used for authentication or no authentication information was provided. |
-| invalid_grant           | The *code* or *refresh_token* value was invalid, or the *redirect_uri* does not match the one used in the authorization request, or the resource owner password is wrong.                                                                      |
+| invalid_grant           | The *code*, *refresh_token*,or user crednetials are invalid, or the *redirect_uri* does not match the one used in the authorization request, or the resource owner password is wrong.                                                                      |
```
user credentials is not needed since it is actually the resource owner password.

mystiberry-okta replied:
One edit requested it, that's why I added it. Will remove again.

yuliu-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -867,30 +867,36 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+refresh_token      | Required if `grant_type` is `refresh_token`. The value is what was returned from this endpoint via a previous invocation. | String |
+username           | Required if the grant_type is `password`. | String |
+password           | Required if the grant_type is `password`. | String |
+scope              | Required if `refresh_token`, or if `password` is the `grant_type`. This is a list of scopes that the client wants to be included in the Access Token. For the *refresh_token* grant type, these scopes have to be subset of the scopes used to generate the Refresh Token in the first place. | String |
```
It is not required for refresh_token grant_type

mystiberry-okta replied:
done.

yuliu-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -188,46 +187,46 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
 
-##### Scope and Grant Type Parameter Details
-
-* If the *grant_type* is `refresh_token` and *scope* is specified, you must specify a subset of the scopes used to generate the Refresh Token in the first place.
```
why remove L215 to L218?

mystiberry-okta replied:
I put all the information in the Response Parameter table, because it was scattered in three different places.  I wasn't sure if this applied to any token or just Refresh Token  (line 223).

yuliu-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -188,46 +187,46 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+| username      | Required if the *grant_type* is `password`.                                                                                                                                                                                                                                                                              | String |
+| password      | Required if the *grant_type* is `password`.                                                                                                                                                                                                                                                                              | String |
+| scope         | Optional.          [ Different scopes and tokens](#response-parameters) are returned depending on the values of `scope` and `grant_type`.                                                                                                                                                                                         | String |
+| redirect_uri  | Required if *grant_type* is `authorization_code`. Specifies the callback location where the authorization was sent; must match what is preregistered in Okta for this client.                                                                                                                                            | String |
```
must match the redirect_uri used to generate the auhtorization_code in the first place.

mystiberry-okta replied:
done.

yuliu-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -867,30 +867,36 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+username           | Required if the grant_type is `password`. | String |
+password           | Required if the grant_type is `password`. | String |
+scope              | Required if `refresh_token`, or if `password` is the `grant_type`. This is a list of scopes that the client wants to be included in the Access Token. For the *refresh_token* grant type, these scopes have to be subset of the scopes used to generate the Refresh Token in the first place. | String |
+redirect_uri       | Required if `grant_type` is `authorization_code`. Specifies the callback location where the authorization was sent; must match what is registered in Okta for this client. | String |
```
must match the redirect_uri used to generate the auhtorization_code in the first place.

mystiberry-okta replied:
Done.

yuliu-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -867,30 +867,36 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+password           | Required if the grant_type is `password`. | String |
+scope              | Required if `refresh_token`, or if `password` is the `grant_type`. This is a list of scopes that the client wants to be included in the Access Token. For the *refresh_token* grant type, these scopes have to be subset of the scopes used to generate the Refresh Token in the first place. | String |
+redirect_uri       | Required if `grant_type` is `authorization_code`. Specifies the callback location where the authorization was sent; must match what is registered in Okta for this client. | String |
+code_verifier      | The code verifier of [PKCE](#parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request. | String |
```
The description for code_verifier, client_id and client_secret should be the same as in the table for OAuth2

mystiberry-okta replied:
done.

yuliu-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -903,31 +909,34 @@ You can't change this value in the Okta user interface.
-password           | Access Token, Refresh Token           |
-client_credentials | Access Token                          |
+| Requested grant type | Requested scope                               | Response tokens                       |
+|:---------------------|:----------------------------------------------|:--------------------------------------|
```
same comments as above grant_type, scopes table

mystiberry-okta replied:
done.

yuliu-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -903,31 +909,34 @@ You can't change this value in the Okta user interface.
-unsupported_grant_type  | The grant_type was not *authorization_code* or *refresh_token*. |
+invalid_request         | The request structure was invalid. For example: the basic authentication header is malformed; both header and form parameters were used for authentication; or no authentication information was provided. |
+invalid_grant           | The `code` or `refresh_token` value is invalid, or the `redirect_uri` does not match the one used in the authorization request. |
+unsupported_grant_type  | The `grant_type` isn't `authorization_code` or `refresh_token`. |
```
`authorization_code`, `refresh_token` or `password`

mystiberry-okta replied:
Done.

mystiberry-okta commented:
@yuliu-okta I addressed your comments. Please check the tables of grant/scope/token carefully, I'm not sure I got the logic right. Appreciate your great comments on these. If you think we should just make the general statement instead of using the tables, feel free to suggest that. I just found that there were so many questions that putting it all in a table made it clearer. 

I also moved some items from response section to request section as suggested.

mystiberry-okta commented:
Attached PDF of the two pages, but remember links likely don't work, and there are funky issues with spacing and page breaks sometimes. (save to PDF is fraught with peril). 
[oidc-draft.pdf](https://github.com/okta/okta.github.io/files/1083313/oidc-draft.pdf)
[oauth2-draft.pdf](https://github.com/okta/okta.github.io/files/1083312/oauth2-draft.pdf)

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -867,30 +867,43 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
 {:.api .api-operation}
 
-{% api_operation post /oauth2/v1/token %}
+{% api_operation post /oauth2/:authorizationServerId/v1/token %}
```
This doesn't seem right; it should still be ```/oauth2/v1/token``` as this doc is for OIDC and not OAuth.

mystiberry-okta replied:
Yup! That's how it is in the new postman collection, too. Fixed.

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -867,30 +867,43 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
-scope              | Optional if *refresh_token*, or *password* is specified as the grant type. This is a list of scopes that the client wants to be included in the Access Token. For the *refresh_token* grant type, these scopes have to be subset of the scopes used to generate the Refresh Token in the first place. | String |
-redirect_uri       | Expected if grant_type is *authorization_code*. Specifies the callback location where the authorization was sent; must match what is preregistered in Okta for this client. | String |
-code_verifier      | The code verifier of [PKCE](#parameter-details). Okta uses it to recompute the code_challenge and verify if it matches the original code_challenge in the authorization request. | String |
+grant_type         | Can be one of the following: `authorization_code`, `password`, or `refresh_token`. Determines the mechanism Okta uses to authorize the creation of the tokens. | String |
```
I'm not 100% certain which of these are appropriate to mention;
1. refresh_token: Whether we should include refresh_token as a valid grant type for OIDC. It doesn't return an ID Token even if you requested openid scope initially to obtain the refresh token, however it is valid to use it as a grant type on this endpoint. It is however not part of the OIDC spec.
2. password: This is probably ok, since we've extended our product to return an ID Token using this grant type of you request openid scope even though it is not part of the OIDC spec.
@yuliu-okta to confirm what we want to doc; whether we want to list out the possibilities at the endpoint or whether we want to document how our endpoint works for the OIDC spec.

mystiberry-okta replied:
Leaving this as is for now and filed JIRA to addres Lars' great question.

mystiberry-okta replied:
Added #2 though since it's missing.

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -867,30 +867,43 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+password           | Required if the grant_type is `password`. | String |
+scope              | Required if `password` is the `grant_type`. This is a list of scopes that the client wants to be included in the Access Token. For the `refresh_token` grant type, these scopes have to be subset of the scopes used to generate the Refresh Token in the first place. | String |
+redirect_uri       | Required if `grant_type` is `authorization_code`. Specifies the callback location where the authorization was sent. This value must match the `redirect_uri` used to generate the original `authorization_code`. | String |
+code_verifier      | Required if `grant_type` is `authorization_code`  and `code_challenge` was specified in the original `/authorize` request. The code verifier of               [PKCE](#parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request. | String |
```
I wonder if we can rephrase this sentence to be more specific or clear: ```The code verifier of               [PKCE](#parameter-details).```?

mystiberry-okta replied:
Changed to "This value is the code verifier for [PKCE]<link> and added the missing bit from oauth2: 

 * [Proof Key for Code Exchange](https://tools.ietf.org/html/rfc7636) (PKCE) is a stronger mechanism for binding the authorization code to the client than just a client secret, and prevents [a code interception attack](https://tools.ietf.org/html/rfc7636#section-1) if both the code and the client credentials are intercepted (which can happen on mobile/native devices). The PKCE-enabled client creates a large random string as *code_verifier* and derives *code_challenge* from it using the method specified in *code_challenge_method*.
    Then the client passes the *code_challenge* and *code_challenge_method* in the authorization request for code flow. When a client tries to redeem the code, it must pass the *code_verifer*. Okta recomputes the challenge and returns the requested token only if it matches the *code_challenge* in the original authorization request. When a client, whose *token_endpoint_auth_method* is ``none``, makes a code flow authorization request, *code_challenge* is required.
    Since *code_challenge_method* only supports S256, this means that the value for *code_challenge* must be: `BASE64URL-ENCODE(SHA256(ASCII(*code_verifier*)))`. According to the [PKCE spec](https://tools.ietf.org/html/rfc7636), the *code_verifier* must be at least 43 characters and no more than 128 characters.

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -867,30 +867,43 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+For web and native application types, an additional process is required:
+
+1. Use the Okta Administration UI and check the **Refresh Token** checkbox under **Allowed Grant Types** on the client application page.
+2. Pass the `offline_access` scope to your `/authorize` request.
```
```Pass the `offline_access` scope to your `/authorize` request or `/token` request if you're using `password` grant type.```

mystiberry-okta replied:
done.

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -903,31 +916,28 @@ You can't change this value in the Okta user interface.
-password           | Access Token, Refresh Token           |
-client_credentials | Access Token                          |
+Based on the `grant_type` and sometimes `scope`, the response contains different token sets.
+Generally speaking, the scopes specified in a request are included in the tokens in the response. 
```
```are included in the tokens``` -> ```are included in the access tokens```

mystiberry-okta replied:
done.

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -903,31 +916,28 @@ You can't change this value in the Okta user interface.
-2. Pass the *offline_access* scope to your authorize request.
+| Requested grant type | Requested scope                                | Response tokens                                                   |
+|:---------------------|:-----------------------------------------------|:------------------------------------------------------------------|
+| authorization_code   | No scope                                       | Access Token. Contains scopes requested in `/authorize` endpoint. |
```
The requested scope is no scope, so I don't see the value in saying "Contains scopes requested in `/authorize`" as there are none in this example.

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -903,31 +916,28 @@ You can't change this value in the Okta user interface.
+| authorization_code   | No scope                                       | Access Token. Contains scopes requested in `/authorize` endpoint. |
+| authorization_code   | Any scopes except `offline_access` or `openid` | Access Token                                                      |
+| authorization_code   | Any or no scopes plus `offline_access`         | Access Token, Refresh Token                                       |
+| authorization_code   | Any or no scopes plus `openid`                 | Access Token, ID Token                                            |
```
Can we also add an example with offline_access and openid, to show that you're getting all 3 tokens?

mystiberry-okta replied:
done.

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -903,31 +916,28 @@ You can't change this value in the Okta user interface.
+| refresh_token        | Any scopes except `offline_access`             | Access Token                                                      |
+| refresh_token        | Any or no scopes plus `offline_access`         | Access Token, Refresh Token                                       |
+| password             | Any or no scopes except `offline_access`       | Access Token                                                      |
+| password             | Any or no scopes plus `offline_access`         | Access Token, Refresh Token                                       |
```
We can now also include the openid scope and get an ID Token, so an example similar to authorization_code might be good here.

mystiberry-okta replied:
Done.

larsjohansen-okta commented on _source/_docs/api/resources/oidc.md:
```diff
@@ -903,31 +916,28 @@ You can't change this value in the Okta user interface.
-invalid_grant           | The *code* or *refresh_token* value was invalid, or the *redirect_uri* does not match the one used in the authorization request. |
-unsupported_grant_type  | The grant_type was not *authorization_code* or *refresh_token*. |
+invalid_request         | The request structure was invalid. For example: the basic authentication header is malformed; both header and form parameters were used for authentication; or no authentication information was provided. |
+invalid_grant           | The `code` or `refresh_token` value is invalid, or the `redirect_uri` does not match the one used in the authorization request. |
```
The `code`, `refresh_token` or `username` and `password` combination is invalid, or the `redirect_uri` does not match the one used in the authorization request.

mystiberry-okta replied:
done.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -248,13 +228,32 @@ For web and native application types, an additional process is required:
+{:.api .api-response .api-response-params}
+
+Based on the grant type and in some cases scope specified in the request, the response contains different token sets.
+Generally speaking, the scopes specified in a request are included in the tokens in the response. 
```
``` in the tokens in the response``` -> ``` in the access tokens in the response```

mystiberry-okta replied:
done.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -188,46 +187,27 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+| password      | Required if the *grant_type* is `password`.                                                                                                                                                                                                                                                                               | String |
+| scope         | Optional.            [ Different scopes and tokens](#response-parameters) are returned depending on the values of `scope` and `grant_type`.                                                                                                                                                                                          | String |
+| redirect_uri  | Required if *grant_type* is `authorization_code`. Specifies the callback location where the authorization was sent. This value must match the `redirect_uri` used to generate the original `authorization_code`.                                                                                                          | String |
+| code_verifier | Required if *grant_type* is `authorization_code`  and `code_challenge` was specified in the original `/authorize` request. The code verifier of                 [PKCE](#request-parameter-details). Okta uses it to recompute the `code_challenge` and verify if it matches the original `code_challenge` in the authorization request. | String |
```
Same comment here, I think we could improve this sentence: ```The code verifier of                 [PKCE](#request-parameter-details).```.

mystiberry-okta replied:
changes the same as oidc above.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -188,46 +187,27 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+
+>Note: You can't provide a `client_id` in both the Authorization header and as a request parameter.
 
 ##### Token Authentication Method
```
Why is this section and the next swapped order in the OIDC doc?

larsjohansen-okta replied:
Also, ```Token Authentication Method``` and ```Refresh Tokens for Web and Native Applications``` could be the same for both of the docs, and the rewrite for the OIDC doc is pretty good, can't we also rewrite it here?

mystiberry-okta replied:
Done!

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -188,46 +187,27 @@ http://www.example.com/#error=invalid_scope&error_description=The+requested+scop
+| client_id     | Required if client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_secret` to authenticate the client application.                                                                                                               | String |
+| client_secret | Required if the client has a secret and client credentials are not provided in the Authorization header. This is used in conjunction with `client_id` to authenticate the client application.                                                                                                                 | String |
+
+>Note: You can't provide a `client_id` in both the Authorization header and as a request parameter.
```
This seems redundant if you rewrite the below two sections as done in the OIDC doc.

mystiberry-okta replied:
We've had support cases on this point. I'd like to leave it in.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -248,13 +228,32 @@ For web and native application types, an additional process is required:
+| Requested grant type | Requested scope                                                            | Tokens in the response                                                |
+|:---------------------|:---------------------------------------------------------------------------|:----------------------------------------------------------------------|
+| authorization_code   | None                                                                       | Access Token. Contains scopes requested in the `/authorize` endpoint. |
+| authorization_code   | `offline_scope`                                                            | Access Token, Refresh Token                                           |
```
```offline_scope``` -> ```offline_access``` Why is this slightly different from the OIDC doc, here it just says ```offline_scope``` and in the OIDC doc it is more elaborate and says ```Any or no scopes plus
offline_access```? I think we should mirror them all here.

mystiberry-okta replied:
Done.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -248,13 +228,32 @@ For web and native application types, an additional process is required:
+| authorization_code   | None                                                                       | Access Token. Contains scopes requested in the `/authorize` endpoint. |
+| authorization_code   | `offline_scope`                                                            | Access Token, Refresh Token                                           |
+| authorization_code   | `openid`                                                                   | Access Token, ID Token                                                |
+| refresh_token        | None                                                                       | Access Token. Contains scopes used to generate the Refresh Token.     |
```
I think this is a great scenario which we should add to the OIDC table.

larsjohansen-okta replied:
This will also return Refresh Token as offline_access scope was necessarily requested in the request that generated the refresh token grant.

mystiberry-okta replied:
Done for both comments.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -248,13 +228,32 @@ For web and native application types, an additional process is required:
+| authorization_code   | `offline_scope`                                                            | Access Token, Refresh Token                                           |
+| authorization_code   | `openid`                                                                   | Access Token, ID Token                                                |
+| refresh_token        | None                                                                       | Access Token. Contains scopes used to generate the Refresh Token.     |
+| refresh_token        | Subset of scopes used to generate Refresh Token                            | Access Token. Contains specified scopes.                              |
```
Subset of scopes used to generate Refresh Token excluding `offline_scope`.

mystiberry-okta replied:
Done.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -248,13 +228,32 @@ For web and native application types, an additional process is required:
+| authorization_code   | `openid`                                                                   | Access Token, ID Token                                                |
+| refresh_token        | None                                                                       | Access Token. Contains scopes used to generate the Refresh Token.     |
+| refresh_token        | Subset of scopes used to generate Refresh Token                            | Access Token. Contains specified scopes.                              |
+| refresh_token        | `offline_scope`                                                            | Access Token, Refresh Token                                           |
```
Subset of scopes used to generate Refresh Token including `offline_access`.

mystiberry-okta replied:
done.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -248,13 +228,32 @@ For web and native application types, an additional process is required:
+| password             | None                                                                       | Access Token. Contains default scopes granted by policy.              |
+| password             | `offline_scope`                                                            | Access Token, Refresh Token. Contains specified scopes.               |
+| password             | `openid`                                                                   | Access Token, ID Token                                                |
+| client_credentials   | Any or no scope                                                            | Access Token. Contains default scopes granted by policy.              |
```
No scope is not allowed, the request has to include at least one scope.

larsjohansen-okta replied:
```Contains default scopes granted by policy in addition to requested scopes.```

mystiberry-okta replied:
Done.

larsjohansen-okta commented on _source/_docs/api/resources/oauth2.md:
```diff
@@ -248,13 +228,32 @@ For web and native application types, an additional process is required:
 | invalid_client          | The specified client ID wasn't found or authentication failed.                                                                                                                                                                 |
 | invalid_request         | The request structure was invalid. E.g. the basic authentication header was malformed, or both header and form parameters were used for authentication or no authentication information was provided. |
-| invalid_grant           | The *code* or *refresh_token* value was invalid, or the *redirect_uri* does not match the one used in the authorization request, or the resource owner password is wrong.                                                                      |
+| invalid_grant           | The *code* or *refresh_token* are invalid, or the *redirect_uri* does not match the one used in the authorization request, or the resource owner password is wrong.                                                                      |
```
The `code`, `refresh_token` or `username` and `password` combination is invalid, or the redirect_uri does not match the one used in the authorization request. (<- to keep it consistent with the OIDC doc)

mystiberry-okta replied:
Done.

mystiberry-okta commented:
Updated PDFs
[oidc-draft.pdf](https://github.com/okta/okta.github.io/files/1083441/oidc-draft.pdf)
[oauth2-draft.pdf](https://github.com/okta/okta.github.io/files/1083440/oauth2-draft.pdf)
